lib/proxy/service.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"net"
+	"strings"
+
+	"github.com/gravitational/teleport/api/client/proto"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+// proxyService impelements the grpc ProxyService.
+type proxyService struct {
+	clusterDialer ClusterDialer
+	log           logrus.FieldLogger
+}
+
+// DialNode opens a bidrectional stream to the requested node.
+func (s *proxyService) DialNode(stream proto.ProxyService_DialNodeServer) error {
+	frame, err := stream.Recv()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// The first frame is always expected to be a dial request.
+	dial := frame.GetDialRequest()
+	if dial == nil {
+		return trace.BadParameter("invalid dial request")
+	}
+
+	if dial.Source == nil || dial.Destination == nil {
+		return trace.BadParameter("invalid dial request")
+	}
+
+	log := s.log.WithFields(logrus.Fields{
+		"node": dial.NodeID,
+		"src":  dial.Source.Addr,
+		"dst":  dial.Destination.Addr,
+	})
+	log.Debugf("Dial request from peer.")
+
+	_, clusterName, err := splitServerID(dial.NodeID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	source := &utils.NetAddr{
+		Addr:        dial.Source.Addr,
+		AddrNetwork: dial.Source.Network,
+	}
+	destination := &utils.NetAddr{
+		Addr:        dial.Destination.Addr,
+		AddrNetwork: dial.Destination.Network,
+	}
+
+	nodeConn, err := s.clusterDialer.Dial(clusterName, reversetunnel.DialParams{
+		From:     source,
+		To:       destination,
+		ServerID: dial.NodeID,
+		ConnType: dial.TunnelType,
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	streamConn := newStreamConn(stream, source, destination)
+	go streamConn.start()
+
+	sent, received := pipeConn(stream.Context(), streamConn, nodeConn)
+	log.Debugf("Closing dial request from peer. sent: %d reveived %d", sent, received)
+	return nil
+}
+
+// splitServerID splits a server id in to a node id and cluster name.
+func splitServerID(address string) (string, string, error) {
+	split := strings.Split(address, ".")
+	if len(split) != 2 {
```

Comment:
Does this assume that cluster names can't have dots?

Reply:
Fixed here [4cb4bfc](https://github.com/gravitational/teleport/pull/10223/commits/4cb4bfcff55cb9cb6fc30a336e13a29c0618cccd)